### ISSUE TITLE
[web-animations-2] Web animations API for specifying timeline phases in animation options.

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#7589]
As part of #7044 we [resolved](https://github.com/w3c/csswg-drafts/issues/7044#issuecomment-1175722383):
* Add ability to specify start and end delays in terms of position in a named phase position (keyword + percentage), e.g. animation-delay: contain 0%; animation-duration: contain 100% attaches the animation between the 0% and 100% points in the contain phase. (See also https://github.com/w3c/csswg-drafts/issues/4342].)
* Add shorthand that expands a phase name to 0% start delay and 100% end delay. E.g. animation-phase: contain expands to animation-delay: contain 0%; animation-end-delay: contain 100%.

These resolutions specify how in CSS you can target an animation to a particular phase of the view timeline. However, we also need a way to do the same from the web animations API. In particular the options passed for the timing of web animations are [EffectTiming](https://www.w3.org/TR/web-animations-1/#dictdef-effecttiming) and [OptionalEffectTiming](https://www.w3.org/TR/web-animations-1/#dictdef-optionaleffecttiming).

Here are a few options I can think of:
1. The simplest would be to extend the `delay` and `endDelay` options to also accept a string specification equivalent to the CSS counterpart (e.g. `delay: "enter 20%"`).
2. Extend `delay` and `endDelay` to also accept an object type representing the combination of phase and time, e.g. `delay: {phase: "enter", offset: CSS.percentage(20)}`.
3. Add new options, e.g. `startTime` / `endTime` or `phase` for specifying the "phase", "time" pairs. We would then have to decide what to do when `delay` and `endDelay` are also specified.

**Comment by kevers-google** [comment]
Leaning towards option 1 as it seems easier for the web developer (less verbose than option 2 and avoids potential confusion due to over-specification with option 3).

**Comment by ydaniv** [comment]
Also leaning towards 1, as authors expect the css animations and web animations APIs to reflect each other as much as possible.

**Comment by flackr** [comment]
Option 1 certainly seems the most straightforward, though option 2 is more consistent with the direction we've been trying to move in terms of having full rich types for the things that are expressed as strings in CSS.

Regardless of option (1) or (2), we will still have similar over-specification issues if we also allow specifying non-"auto" durations e.g. #4342. I think we'll just have to choose a precedent - probably try to align this with what happens when you specify left, right and width on an element size.

**Comment by flackr** [comment]
@birtles @graouts FYI

**Comment by flackr** [comment]
Another variant of 2 would be a new unit type to represent the phase + percentage pair, e.g. `delay: CSS.PhasePercent('enter', 20)` - though probably not under CSS given it's not a CSSNumber?

**Comment by birtles** [comment]
Option 1 seems fine to me. I _really_ want to avoid having `phase` as a mutable property ([1](https://github.com/w3c/csswg-drafts/issues/7044#issuecomment-1060133970), [2](https://github.com/w3c/csswg-drafts/issues/7044#issuecomment-1145569285)).

We could possibly introduce option 2 (or the variant you mentioned) at a later point if it proves necessary?

**Comment by flackr** [comment]
> I _really_ want to avoid having `phase` as a mutable property ([1](https://github.com/w3c/csswg-drafts/issues/7044#issuecomment-1060133970), [2](https://github.com/w3c/csswg-drafts/issues/7044#issuecomment-1145569285)).

Yeah I'm hoping to bikeshed that on #7044 or possibly open a specific issue to bikeshed it. I've added a comment there.

**Comment by flackr** [comment]
The one downside to starting with option 1 is that if this is the string value is what you read back from [getTiming](https://drafts.csswg.org/web-animations-1/#dom-animationeffect-gettiming) then we can't upgrade to option 2 without risking breaking content.

**Comment by birtles** [comment]
> The one downside to starting with option 1 is that if this is the string value is what you read back from [getTiming](https://drafts.csswg.org/web-animations-1/#dom-animationeffect-gettiming) then we can't upgrade to option 2 without risking breaking content.

Yeah, I mean we _could_ introduce parallel properties -- typed and untyped/string-type -- and precedence rules for when they're both set, but then we'd break code that attempts to roundtrip the values, by fetching the timing, setting only the lower precedence one (e.g. the untyped/string-type one), then re-setting it.

**Comment by flackr** [comment]
My strawman proposal for the object would be something like the following:
```
dictionary TimelineRangeOffset {
  CSSOMString? rangeName;
  CSSNumericValue offset;
}
```

The name TimelineRangeOffset is based on the term name from #7701 but this could also be AnimationDelay, TimelineRange, or something else more appropriate.

Using a CSSNumericValue would in theory allow for arbitrary offsets from the start / end of a range, e.g. `calc(100% + 10px)`. Also, without a range name specified this would allow specifying offsets such as CSS.px(50) (e.g. which should support #7575 #7296 #4342).

**Comment by fantasai** [comment]
> Note that in the context of keyframe offsets https://github.com/w3c/csswg-drafts/issues/7637 the offset name does feel a bit redundant, e.g.

Could you have the API automatically wrap a raw CSSNumericValue around a TimelineOffset with a null rangeName?

**Comment by flackr** [comment]
> Could you have the API automatically wrap a raw CSSNumericValue around a TimelineOffset with a null rangeName?

If I'm following correctly, the idea is that if you specify a TimelineOffset with null rangeName it would effectively use that TimelineOffset's inner `offset` value? Seems reasonable.

**Comment by fantasai** [comment]
@flackr No, what I was wondering was if we could collapse your example:

```
element.animate([
  {opacity: 0, offset: { offset: CSS.percent(20) }},
  {opacity: 1, offset: { offset: CSS.percent(80) }}], {
    delay: { rangeName: 'enter', offset: CSS.percent(0) },
    endDelay: { rangeName: 'enter', offset: CSS.percent(100) },
  });
```
  
  down to 

```
element.animate([
  {opacity: 0, offset:CSS.percent(20)},
  {opacity: 1, offset: CSS.percent(80)}], {
    delay: { rangeName: 'enter', offset: CSS.percent(0) },
    endDelay: { rangeName: 'enter', offset: CSS.percent(100) },
  });
```

(I'm not familiar with WebIDL and JS, so my main frame of reference is C++, thus all the stupid questions...)

**Comment by flackr** [comment]
Gotcha, I think this would just be accepting `TimelineRangeOffset or CSSNumberish` and then when we parse the arguments in C++ we could determine the type of the passed value.

**Comment by flackr** [comment]
Based on the resolution in #7901 animation-range should be set with a new set of properties. As a new set of properties we don't have to support legacy values. Based on the resolution in #7748 we should aim for accepting a string equivalent to the CSS syntax or typed OM. Rough proposal, add the following to [EffectTiming](https://www.w3.org/TR/web-animations-1/#dictdef-effecttiming) and [OptionalEffectTiming](https://www.w3.org/TR/web-animations-1/#dictdef-optionaleffecttiming):

```
dictionary TimelineRangeOffset {
  CSSOMString? rangeName;
  CSSNumericValue offset;
}
DOMString or TimelineRangeOffset rangeStart = "auto";
DOMString or TimelineRangeOffset rangeEnd = "auto";
```

DOMStrings would be parsed the same as the CSS animation-range-start and animation-range-end properties, as `<timeline-range-name>? <percentage> | "auto"`.

**Comment by birtles** [comment]
I thought the idea was that animation-range would be a property of the animation, not the effect?

**Comment by birtles** [comment]
I think the question is, what should happen with regard to nested effects like `GroupEffect` and `SequenceEffect`?

I think it makes sense that you set up the animation range on the root `Animation` (since the `Animation` is the thing that binds us to a timeline and a range is specific to a timeline) and the result of that is applied to the `Animation`'s root effect.

**Comment by flackr** [comment]
Yeah that's a good point. In the context of `GroupEffect` and `SequenceEffect` it probably makes more sense to put on the `Animation` so that it can set the range for the root effect and result in the group or sequence covering that range. So the proposal is then https://github.com/w3c/csswg-drafts/issues/7589#issuecomment-1380842759 but added to [KeyframeAnimationOptions](https://www.w3.org/TR/web-animations-1/#dictdef-keyframeanimationoptions).

**Comment by flackr** [comment]
I believe this is resolved now with 52b620ffd4364680eedf23cb72738dee671ef44c landed.
